contracts/deposit-contracts/contract/src/detail.rs

```diff
@@ -0,0 +1,110 @@
+/*
+    The utilities found in this file were scraped from other Casper contracts,
+    mainly cep-78 and cep-18.
+    This file is not necessarily due for review, unless breaking changes are suspected.
+*/
+
+use crate::error::DepositError;
+use alloc::vec::Vec;
+use casper_contract::{
+    contract_api::{self, runtime},
+    ext_ffi,
+    unwrap_or_revert::UnwrapOrRevert,
+};
+use casper_types::{
+    api_error, bytesrepr, bytesrepr::FromBytes, system::CallStackElement, ApiError, Key, URef,
+};
+use core::convert::TryInto;
+
+fn call_stack_element_to_address(call_stack_element: CallStackElement) -> Key {
+    match call_stack_element {
+        CallStackElement::Session { account_hash } => Key::from(account_hash),
+        CallStackElement::StoredSession { account_hash, .. } => {
+            // Stored session code acts in account's context, so if stored session wants to interact
+            // with an CEP-18 token caller's address will be used.
+            Key::from(account_hash)
+        }
+        CallStackElement::StoredContract {
+            contract_package_hash,
+            ..
+        } => Key::from(contract_package_hash),
+    }
+}
+
+/// Traverse the callstack
+pub(crate) fn get_immediate_caller() -> Result<Key, DepositError> {
+    let call_stack = runtime::get_call_stack();
+    call_stack
+        .into_iter()
+        .rev()
+        .nth(1)
+        .map(call_stack_element_to_address)
+        .ok_or(DepositError::InvalidContext)
+}
+
+/// Gets [`URef`] under a name.
+pub(crate) fn get_uref(name: &str) -> URef {
+    let key = runtime::get_key(name)
+        .ok_or(ApiError::MissingKey)
+        .unwrap_or_revert();
+    key.try_into().unwrap_or_revert()
+}
+
+pub fn get_named_arg_size(name: &str) -> Option<usize> {
+    let mut arg_size: usize = 0;
+    let ret = unsafe {
+        ext_ffi::casper_get_named_arg_size(
+            name.as_bytes().as_ptr(),
+            name.len(),
+            &mut arg_size as *mut usize,
+        )
+    };
+    match api_error::result_from(ret) {
+        Ok(_) => Some(arg_size),
+        Err(ApiError::MissingArgument) => None,
+        Err(e) => runtime::revert(e),
+    }
+}
+
+/// Optional named args are used for access control
+pub fn get_optional_named_arg_with_user_errors<T: FromBytes>(
+    name: &str,
+    invalid: DepositError,
+) -> Option<T> {
+    match get_named_arg_with_user_errors::<T>(name, DepositError::Phantom, invalid) {
+        Ok(val) => Some(val),
+        Err(_) => None,
+    }
+}
+
```

Comment:
What I meant by the comment at the top of the detail.rs file is that I am making use of these functions because they were used in cep contracts. I didn't not write them and have no intention to touch them unless something is wrong / causing a problem.

Reply:
I don't understand why `get_optional_named_arg_with_user_errors()` is useful.

Besides, there are many concerns about CEP-78 quality, so I would not just copy their code without further considerations.

Reply:
Notice that error is not used at all - you could replace `invalid` with `DepositError::Phantom` and it would still work. They got many things wrong, even simple things like getting optional args.

Reply:
I have changed this code since, therefore I am resolving this conversation since it points to an out-of-date commit.